### PR TITLE
fix(47508): noUncheckedIndexedAccess with enums Type narrowed

### LINKED ISSUE #47508: noUncheckedIndexedAccess with enums could be type narrowed
# Bug Report

Hi, we're trying to tighten up some of our code by using `noUncheckedIndexedAccess`. I know there are caveats to this compiler option which we've found by searching, but this one feels more like a bug?

### 🔎 Search Terms

Searched a variety of issues, the most relevant being: https://github.com/microsoft/TypeScript/issues/13778

### 🕗 Version & Regression Information

`4.5.4`

### 💻 Code

```ts 
    enum Meat {
      Sausage,
      Bacon
    }
    const sausage = Meat.Sausage
    const value = Meat[sausage] // string | undefined
``` 

### 🙁 Actual behavior

`value` in the above example can't be undefined

### 🙂 Expected behavior

`value` to be typed as string, not `string | undefined`

### PR DESCRIPTION
**Pull request by navya9singh**
Fixes #47508

### CODE REVIEW

**Comment by navya9singh** [comment]
> This PR doesn't have any linked issues. Please open an issue that references this PR. From there we can discuss and prioritise.

https://github.com/microsoft/TypeScript/issues/47508

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -15589,6 +15589,11 @@ m2: ${(this.mapper2 as unknown as DebugTypeMapper).__debugToString().split("\n")
                         return accessFlags & AccessFlags.IncludeUndefined ? getUnionType([indexInfo.type, undefinedType]) : indexInfo.type;
                     }
                     errorIfWritingToReadonlyIndex(indexInfo);
+                    if (accessFlags & AccessFlags.IncludeUndefined && objectType.symbol.flags & (SymbolFlags.RegularEnum | SymbolFlags.ConstEnum)) {
+                        if(indexType.flags & TypeFlags.EnumLiteral){
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts:15593]:
```suggestion
                        if (indexType.flags & TypeFlags.EnumLiteral) {

```

**Comment by RyanCavanaugh** [comment]
It looks like this also needs to check that the indexing enum type is the same as the indexed enum type, in other words this example should not have its behavior changed:
```ts
enum A {
  a, b, c
}
enum B {
  x, y, z
}
const p = A[B.x];
```

File: src/compiler/checker.ts (hunk 1 above)
Comment (ProKashif) [reply src/compiler/checker.ts:15593]: +1

File: tests/baselines/reference/noUncheckedIndexAccess.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  const union: Meat.Bacon | Meat.Sausage = Meat.Bacon
+  const valueUnion = Meat[union]
+
+  //Avoiding a false positive
+  const value = Meat[0]
+
+  const t = "testing"
+  const value2 = Meat[t]
+
```
Comment (ProKashif) [inline tests/baselines/reference/noUncheckedIndexAccess.js:20]:
```suggestion

```

File: tests/cases/compiler/noUncheckedIndexAccess.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  const valueBacon = Meat[bacon]
+
+  const union: Meat.Bacon | Meat.Sausage = Meat.Bacon
+  const valueUnion = Meat[union]
+
+  //Avoiding a false positive
+  const value = Meat[0]
+
+  const t = "testing"
+  const value2 = Meat[t]
```
Comment (ProKashif) [inline tests/cases/compiler/noUncheckedIndexAccess.ts:21]: `t` variable name should be self-explaining

**Review by ProKashif — CHANGES_REQUESTED** [review]

File: src/compiler/checker.ts (hunk 1 above)
Comment (navya9singh) [reply src/compiler/checker.ts:15593]: Made some changes, please have a look!

File: tests/cases/compiler/noUncheckedIndexAccess.ts (hunk 3 above)
Comment (navya9singh) [reply tests/cases/compiler/noUncheckedIndexAccess.ts]: Made some changes, please have a look!

**Review by sandersn — APPROVED** [review]

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -15589,6 +15589,11 @@ m2: ${(this.mapper2 as unknown as DebugTypeMapper).__debugToString().split("\n")
                         return accessFlags & AccessFlags.IncludeUndefined ? getUnionType([indexInfo.type, undefinedType]) : indexInfo.type;
                     }
                     errorIfWritingToReadonlyIndex(indexInfo);
+                    if (accessFlags & AccessFlags.IncludeUndefined && objectType.symbol && objectType.symbol.flags & (SymbolFlags.RegularEnum | SymbolFlags.ConstEnum)) {
+                        if (indexType.symbol && indexType.flags & TypeFlags.EnumLiteral && getParentOfSymbol(indexType.symbol) === objectType.symbol) {
+                            return indexInfo.type;
+                        }
+                    }
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts:15596]: Both the initial test `accessFlags & AccessFlags.IncludeUndefined` and one of the possible outcomes `indexInfo.type` are shared between the inserted block and the final `return`. This doesn't need to be a giant ternary but all the logic should be folded into a coherent control flow structure here.

Comment (navya9singh) [reply src/compiler/checker.ts:15596]: Made some changes here to fix the flow.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -15589,7 +15589,13 @@ m2: ${(this.mapper2 as unknown as DebugTypeMapper).__debugToString().split("\n")
                         return accessFlags & AccessFlags.IncludeUndefined ? getUnionType([indexInfo.type, undefinedType]) : indexInfo.type;
                     }
                     errorIfWritingToReadonlyIndex(indexInfo);
-                    return accessFlags & AccessFlags.IncludeUndefined ? getUnionType([indexInfo.type, undefinedType]) : indexInfo.type;
+                    if (accessFlags & AccessFlags.IncludeUndefined) {
+                        if (objectType.symbol && objectType.symbol.flags & (SymbolFlags.RegularEnum | SymbolFlags.ConstEnum) && (indexType.symbol && indexType.flags & TypeFlags.EnumLiteral && getParentOfSymbol(indexType.symbol) === objectType.symbol)) {
+                            return indexInfo.type;
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts:15594]: This is still duplicating the line that we started with.

Proposed code (I haven't tested this):
```ts
                    // When accessing an enum object with its own type,
                    // e.g. E[E.A] for enum E { A }, undefined shouldn't
                    // be included in the result type
                    if ((accessFlags & AccessFlags.IncludeUndefined) &&
                        !(objectType.symbol &&
                            objectType.symbol.flags & (SymbolFlags.RegularEnum | SymbolFlags.ConstEnum) &&
                            (indexType.symbol &&
                            indexType.flags & TypeFlags.EnumLiteral &&
                            getParentOfSymbol(indexType.symbol) === objectType.symbol))) {
                        return getUnionType([indexInfo.type, undefinedType]);
                    }
                    return indexInfo.type;
```

**Review by RyanCavanaugh — APPROVED** [review]

**Pull request merged**